sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -913,16 +985,55 @@ public void setValueDeserializer(String valueDeserializer) {
       return input.getPipeline().apply(transform);
     }
 
-    private Coder<K> getKeyCoder(CoderRegistry coderRegistry) {
-      return (getKeyCoder() != null)
-          ? getKeyCoder()
-          : getKeyDeserializerProvider().getCoder(coderRegistry);
+    private Coder<K> getKeyCoder(CoderRegistry coderRegistry, SchemaRegistry schemaRegistry) {
+      final Coder<K> keyCoder =
+          (getKeyCoder() != null)
+              ? getKeyCoder()
+              : getKeyDeserializerProvider().getCoder(coderRegistry);
+      if (isKeyInferBeamSchema()) {
+        return inferSchemaCoder(keyCoder, schemaRegistry);
+      }
+      return keyCoder;
+    }
+
+    private Coder<V> getValueCoder(CoderRegistry coderRegistry, SchemaRegistry schemaRegistry) {
+      final Coder<V> valueCoder =
+          (getValueCoder() != null)
+              ? getValueCoder()
+              : getValueDeserializerProvider().getCoder(coderRegistry);
+      if (isValueInferBeamSchema()) {
+        return inferSchemaCoder(valueCoder, schemaRegistry);
+      }
+      return valueCoder;
+    }
+
+    @VisibleForTesting
+    static <T> SchemaCoder<T> inferSchemaCoder(Coder<T> coder, SchemaRegistry schemaRegistry) {
+      // At the moment there is not a general way to infer SchemaCoders from non-schema Coders (that
```

Comment:
@reuvenlax I tried to get the Avro resolution by registering `AvroRecordSchema` as a a new `SchemaRegistryProvider` but this produced a lot of issues. 

I have the impression we need to do something like what I mention to register which basic Coders we can support also as SchemaCoders e.g. Avro, Protobuf and others that contain a `Schema` and the `from/toRowFunction`s